apis/gateway/v1beta2/apirule_types.go

```diff
@@ -0,0 +1,236 @@
+)
+
+// APIRuleSpec defines the desired state of ApiRule.
+type APIRuleSpec struct {
+	// Specifies the URL of the exposed service.
+	// +kubebuilder:validation:MinItems=1
+	Hosts []*string `json:"hosts"`
+	// Describes the service to expose.
+	// +optional
+	Service *Service `json:"service,omitempty"`
+	// Specifies the Istio Gateway to be used.
+	// +kubebuilder:validation:Pattern=`^[0-9a-z-_]+(\/[0-9a-z-_]+|(\.[0-9a-z-_]+)*)$`
+	Gateway *string `json:"gateway"`
+	// Specifies CORS headers configuration that will be sent downstream
+	// +optional
+	CorsPolicy *CorsPolicy `json:"corsPolicy,omitempty"`
+	// Represents the array of Oathkeeper access rules to be applied.
+	// +kubebuilder:validation:MinItems=1
+	Rules []Rule `json:"rules"`
+	// +optional
+	Timeout *Timeout `json:"timeout,omitempty"`
+}
+
+// APIRuleStatus describes the observed state of ApiRule.
+type APIRuleStatus struct {
+	LastProcessedTime    *metav1.Time           `json:"lastProcessedTime,omitempty"`
+	ObservedGeneration   int64                  `json:"observedGeneration,omitempty"`
+	APIRuleStatus        *APIRuleResourceStatus `json:"APIRuleStatus,omitempty"`
+	VirtualServiceStatus *APIRuleResourceStatus `json:"virtualServiceStatus,omitempty"`
+	// +optional
+	AccessRuleStatus *APIRuleResourceStatus `json:"accessRuleStatus,omitempty"`
+	// +optional
+	RequestAuthenticationStatus *APIRuleResourceStatus `json:"requestAuthenticationStatus,omitempty"`
+	// +optional
+	AuthorizationPolicyStatus *APIRuleResourceStatus `json:"authorizationPolicyStatus,omitempty"`
+}
+
+// APIRule is the Schema for ApiRule APIs.
+// +kubebuilder:object:root=true
+// +kubebuilder:subresource:status
+// +kubebuilder:printcolumn:name="Status",type="string",JSONPath=".status.APIRuleStatus.code"
+// +kubebuilder:printcolumn:name="Host",type="string",JSONPath=".spec.host"
+type APIRule struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   APIRuleSpec   `json:"spec,omitempty"`
+	Status APIRuleStatus `json:"status,omitempty"`
+}
+
+// +kubebuilder:object:root=true
+
+// APIRuleList contains a list of ApiRule
+type APIRuleList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+	Items           []APIRule `json:"items"`
+}
+
+// Service .
+type Service struct {
+	// Specifies the name of the exposed service.
+	Name *string `json:"name"`
+	// Specifies the Namespace of the exposed service. If not defined, it defaults to the APIRule Namespace.
+	// +kubebuilder:validation:Pattern=^[a-z0-9]([-a-z0-9]*[a-z0-9])?$
+	// +optional
+	Namespace *string `json:"namespace,omitempty"`
+	// Specifies the communication port of the exposed service.
+	// +kubebuilder:validation:Minimum=1
+	// +kubebuilder:validation:Maximum=65535
+	Port *uint32 `json:"port"`
+	// Specifies if the service is internal (in cluster) or external.
+	// +optional
+	IsExternal *bool `json:"external,omitempty"`
+}
+
+// Rule .
+type Rule struct {
```

Comment:
We missed to add external authorizers.

Reply:
We do not add them since it won't be implemented with this PR. @strekm right?

Reply:
Ah yes, sorry ... my mistake 🙄